### DISCUSSION TITLE
Port your program parse_menu.py for the tinySA

### DISCUSSION THREAD

**Discussion by rudi48** [Ho-Ro/NanoVNA-D#3] [Ideas]
Hello Martin,

Thank you very much for your detailed explanations.

I have tried to port your program parse_menu.py for the tinySA.
Links:
File ui_sa.c and plot.c from:
https://github.com/erikkaashoek/tinySA

The adopted program is named: Ho-Ro_tinySA_menu_readout.py
I added a statement to cut off any // C comments at the end of a line.
It works up to the point:
...
    PAUSE/RESUME SWEEP
    CALC
    STORAGE
static const menuitem_t menu_storage[] = {  -  menu_storage
        STORE TRACE
        CLEAR STORED
        SUBTRACT STORED
        NORMALIZE
        <- BACK
    STORE TRACE
    CLEAR STORED
    SUBTRACT STORED
    NORMALIZE
    WATER FALL
    SWEEP SETTINGS
static const menuitem_t menu_sweep_speed[] =  -  menu_sweep_speed
MT_ADV_CALLBACK,     SD_NORMAL,     "NORMAL",          menu_scanning_speed_acb},    
Traceback (most recent call last):
  File "/Volumes/DAT/HDA8/INSTALL/Develop/Afu/tinySA/Menu/Ho-Ro_tinySA_menu_readout.py", line 149, in <module>
    show_menu( 'menu_top', 0 )
  File "/Volumes/DAT/HDA8/INSTALL/Develop/Afu/tinySA/Menu/Ho-Ro_tinySA_menu_readout.py", line 146, in show_menu
    show_menu( next_menu, level + 1 )
  File "/Volumes/DAT/HDA8/INSTALL/Develop/Afu/tinySA/Menu/Ho-Ro_tinySA_menu_readout.py", line 146, in show_menu
    show_menu( next_menu, level + 1 )
  File "/Volumes/DAT/HDA8/INSTALL/Develop/Afu/tinySA/Menu/Ho-Ro_tinySA_menu_readout.py", line 85, in show_menu
    param = param.strip()
UnboundLocalError: local variable 'param' referenced before assignment

 The last error line comes from the new statements: try: and except:

Now I have no more idea how to make it work.

All files should be attached:
[Ho-Ro_tinySA_menu_readout.py.zip](https://github.com/Ho-Ro/Nano
[plot.c.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9895891/plot.c.zip)
VNA-D/files/9895889/Ho-Ro_tinySA_menu_readout.py.zip)

[ui_sa.c.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9895910/ui_sa.c.zip)

Would you please be so kind and try it make it work?

73, Rudi DL5FA

**Comment by rudi48** [comment]
Hello Martin,
2 files are not accesable:
[Ho-Ro_tinySA_menu_readout.py.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9895933/Ho-Ro_tinySA_menu_readout.py.zip)
[plot.c.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9895935/plot.c.zip)

73, Rudi DL5FA

**Reply by Ho-Ro** [reply]
Hello Rudi,

please check my changed version, just rename from `*.py.txt` to `*.py`:
[Ho-Ro_tinySA_menu_readout.py.txt](https://github.com/Ho-Ro/NanoVNA-D/files/9912068/Ho-Ro_tinySA_menu_readout.py.txt)

I added the `#if 0` handling and changed the c comment processing, your code didn't cut lines, where the `//` is at pos=0 or pos=1.

```diff
--- Ho-Ro_tinySA_menu_readout.py.orig   2022-11-01 16:36:36.046029386 +0100
+++ Ho-Ro_tinySA_menu_readout.py        2022-11-01 17:01:50.073605836 +0100
@@ -33,7 +33,26 @@
 
 # read the file 'ui.c' which contains the menu structure
 with open( 'ui_sa.c' ) as ui_c:
-    ui = ui_c.read().splitlines()
+    lines = ui_c.read().splitlines()
+
+# use line content left of c comment "//"
+pattern_text = r'(?P<line>.*?)(?P<comment>//.*)'
+pattern = re.compile(pattern_text)
+
+skip_lines = False
+ui = []
+
+for line in lines:
+    if '#if 0' in line: # skip the lines starting from here
+        skip_lines = True
+    elif skip_lines:
+        if '#else' in line or '#endif' in line: # stop skipping
+            skip_lines = False
+    else:
+        if '//' in line: # cut comment
+            match = pattern.match( line )
+            line = match.group( 'line' )
+        ui.append( line )
 
 
 # format a string like 'menu_XXXXX_acb' or 'menu_XXXXX_cb' and return only 'XXXXX'
@@ -75,13 +94,7 @@
             or "MT_CALLBACK" in line \
             or "MT_ADV_CALLBACK" in line:
                 line = line.strip(' {},')
-                beginpos = line.find('//')  # RR: remove C comment "//"
-                if beginpos > 1: line = line[0:beginpos]
-                #if line.count(',') > 4: print (line) # RR: debug
-                try:
-                    what, param, text, next_menu = line.split( ',' )
-                except:
-                    print(line)
+                what, param, text, next_menu = line.split( ',' )
                 param = param.strip()
                 text = text.strip()
```
This part of the source code gives an unlimited recursion when I parse ui_sa.c in the cloned repo, `menu_scale_per2` references `menu_scale_per2` which references `menu_scale_per2` which references ... :

```c
#if 0
static const menuitem_t menu_scale_per2[] = {
  { MT_ADV_CALLBACK, 6, "0.1 /",   menu_scale_per_acb},
  { MT_ADV_CALLBACK, 7, "0.2 /",   menu_scale_per_acb},
  { MT_ADV_CALLBACK, 8, "0.05/",   menu_scale_per_acb},
  { MT_ADV_CALLBACK, 9, "0.02/",   menu_scale_per_acb},
  { MT_ADV_CALLBACK,10, "0.01/",   menu_scale_per_acb},
//{ MT_ADV_CALLBACK,11, "0.005/",  menu_scale_per_acb},
//{ MT_SUBMENU,  0, S_RARROW" MORE",   menu_scale_per2},
  { MT_CANCEL,   0, S_LARROW" BACK", NULL },
  { MT_NONE,     0, NULL, NULL } // sentinel
};
```

**Comment by rudi48** [comment]
Hello Manfred,

Thank you very much for your help.
It looks very good in the first view.

73, Rudi DL5FA

**Comment by rudi48** [comment]
Hello Manfred,

Thank you for expanding the program so much.
Motivated by that I added the values for RBW, point, speed and mixer.

Now I found the time to check program Ho-Ro_tinySA_menu_readout.py

Checked every menu block of ui_sa.c, see:
ui_sa.c menu check:
line
1429 store preset high   	OK
1441 load preset high    	OK
1453 lo_drive 				OK
1497 modulation 			?
1509 sweep					?
1520 lowoutputmode			?
1534 highoutputmode			?
1547 average				?
1561 rbw					OK
1584 reffer					?
1597 atten					OK
1606 reflevel				?
1613 search					OK
1662 marker_ops				?
1672 marker					?				
1681 dfu					OK
1688 harmonic				OK
1712 sweep_points			?
1721 sweep_speed			?
1733 settings3				?
1753 settings2				OK
1773 settings				?
1786 measure2				?
1797 measure				OK
1809 calibrate				OK
1831 serial_speed			OK
1845 connection				OK
1854 touch					OK
1861 config					OK
1875 storage				OK
1884 display				OK
1898 unit					OK
1911 trigger				OK
1924 level					OK
1937 stimulus				?
1951 mode					OK
1968 topultra				?
1982 menu_top				OK

menu_stimulus is missing
START
STOP
CENTER
SPAN
ZERO SPAN
RBW

menu_settings is missing
LO OUTPUT      OK
ACTUAL POWER
IF FREQ
SCAN SPEED     OK
SAMPLE REPEAT

...

My question is, is it worth to further work on the question mark points?

73, Rudi DL5FA
[Ho-Ro_tinySA_menu_readout_2022-11-02.py.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9930238/Ho-Ro_tinySA_menu_readout_2022-11-02.py.zip)

[tinySA_menu_map_2022-11-02.txt.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9930459/tinySA_menu_map_2022-11-02.txt.zip)

**Reply by Ho-Ro** [reply]
Hello Rudi,

you could compare your list with this [menu tree](https://tinysa.org/wiki/pmwiki.php?n=Main.MenuTree).
Which FW version does your device show? The tree above is from v1.3-325, the source code is in the [V4.3 branch](https://github.com/erikkaashoek/tinySA/tree/V4.3).

Martin

**Comment by rudi48** [comment]
Hello Martin,

Thank you for taking care.

In the github version the file plot.c and ui_sa.care are just got updatet.
Unfortunately I did not note the date of these files.
I estimate about 2 years old.

The version 1.3.325 is about 1 year old.
I made a hand made menu tree from the tinySA device, see attached ReadMe_tinySA_menu_V1.3-325.txt

OK, now we should have actual files, which I can compare with my tinySA device.
I will repeat my deviation list again, tomorrow.

73, Rudi DL5FA

[ReadMe_tinySA_menu_V1.3-325.txt.zip](https://github.com/Ho-Ro/NanoVNA-D/files/9931847/ReadMe_tinySA_menu_V1.3-325.txt.zip)
